stdlib/hashtbl.ml

```diff
@@ -149,102 +162,82 @@ let resize indexfun h =
   if nsize < Sys.max_array_length then begin
     let ndata = Array.make nsize Empty in
     let inplace = not (ongoing_traversal h) in
-    h.data <- ndata;          (* so that indexfun sees the new bucket count *)
+    h.data <- ndata;  (* so that indexfun sees the new bucket count *)
     insert_all_buckets (indexfun h) inplace odata ndata
   end
 
 let iter f h =
   let rec do_bucket = function
-    | Empty ->
-        ()
-    | Cons{key; data; next} ->
-        f key data; do_bucket next in
-  let old_trav = ongoing_traversal h in
-  if not old_trav then flip_ongoing_traversal h;
-  try
+    | Empty -> ()
+    | Cons {key; data; next} -> f key data; do_bucket next
+  in
+  protect_traversal h (fun () ->
     let d = h.data in
     for i = 0 to Array.length d - 1 do
       do_bucket d.(i)
```

Comment:
Is this just `Array.iter`? If so, we could replace it -- `Array.iter` uses `unsafe_get`, which is nice as well.

Reply:
If you're not using Flambda, then using `Array.iter` will introduce an indirect call. I'm not sure if it matters that much on modern processors (the branch predictor will likely sort it out), but then the same is true of the array bounds check.
So, my guess is that replacing the existing code with `Array.iter do_bucket d` is not particularly interesting for performance.

Reply:
(But it's nice for readability of the code  and its size, so probably a good move if the benchmarks don't say otherwise.)

Reply:
Despite the indirect function call being certainly worse for performance than the bounds check, I'd vote for `Array.iter`.